client/modules/cm-lang-ftml/src/grammars/ftml.yaml

```diff
@@ -0,0 +1,999 @@
+      emit: BlockNode
+      indent: delimited(]])
+      skip: /\s+/
+      chain:
+        - BlockStart
+        - BlockNamePrefix?
+        - BlockNameValueElements
+        - BlockNameSuffix?
+        - BlockValue*
+        - BlockEnd
+    end:
+      type: BlockContainerValueEndNode
+      emit: BlockNode
+      indent: delimited(]])
+      skip: /\s+/
+      chain:
+        - BlockStartClosing
+        - BlockNamePrefix?
+        - BlockNameValueElements
+        - BlockNameSuffix?
+        - BlockEnd
+
+  # -- INCLUDE
+
+  IncludeLink:
+    match: /[^\s\]]+/
+    tag: link
+
+  IncludeNameElements:
+    emit: IncludeName
+    match: "include-elements"
+    lookahead: /@namela/
+    tag: keyword
+
+  IncludeElements:
+    emit: Include
+    indent: delimited(]])
+    skip: /\s+/
+    chain:
+      - BlockStart
+      - BlockNamePrefix?
+      - IncludeNameElements
+      - BlockNameSuffix?
+      - IncludeLink?
+      - BlockNodeArgument |* BlockLabel
+      - BlockEnd
+
+  IncludeNameMessy:
+    emit: IncludeName
+    match: "include-messy"
+    lookahead: /@namela/
+    tag: keyword
+
+  IncludeMessyStart:
+    skip: /\s+/
+    chain:
+      - BlockStart
+      - BlockNamePrefix?
+      - IncludeNameMessy
+      - BlockNameSuffix?
+      - IncludeLink?
+
+  IncludeMessy:
+    begin: IncludeMessyStart
+    end: BlockEnd
+    inside:
+      - type: IncludeMessyArgument
+        begin:
+          match: /([^\s=|]+)(\s*=\s*)/
+          captures:
+            0: { type: IncludeArgumentName, tag: special(propertyName) }
+            1: { is: BlockNodeArgumentOperator }
+        end:
+          match: /@BlockEnd|\|/
+          rematch: true
+        inside:
+          - include: block
+          - include: blocks
+          - include: inline
+      - { type: IncludeMessySeparator, match: "|", tag: separator }
+      - BlockLabel
+
+includes:
+  block:
+    - HorizontalRule
+    - ClearFloat
+    - Heading
+    - Center
+    - Table
+    - Container
+    # - Paragraph
+
+  inline:
+    - Typography
+    - LinkInline
+    - IncludeVariable
+    - PageVariable
+    - RawEscapeBlock
+    - Escaped
+    - ColoredText
+    - Strong
+    - Emphasis
+    - Underline
+    - Strikethrough
+    - Subscript
+    - Superscript
+    - Monospace
+
+  includes:
+    - IncludeElements
+    - IncludeMessy
+
+  blocks:
+    - TripleLink
+    - TripleLinkSimple
+    - BlockNodeInlineMath
+    - BlockNestedMath
+    - BlockNestedModuleCSS
+    - BlockNestedCSS
+    - BlockNestedHTML
+    - BlockNestedCode
+    - BlockNodeMap
+    - BlockNodeValue
+    - BlockNodeValueMap
+    - BlockNodeModule
+    - BlockContainerAlign
+    - BlockContainerElements
+    - BlockContainerMap
+    - BlockContainerValue
+    - BlockNodeUnknown
+    - SingleLink
+    - SingleLinkAnchor
+
+global:
+  - EscapedNewLine
+  - EscapedCharacter
+  - BlockComment
+  - NonControl
+
+root:
+  - include: block
+  - include: includes
+  - include: blocks
+  - include: inline
```

Comment:
I'll say, despite the length, this YAML format is way cleaner to read than the previous version. +1

Reply:
this is like my third attempt at making a grammar definition format

at some point I have to get it right lmao